src/Control/Monad/Free.purs

```diff
@@ -1,245 +1,146 @@
 module Control.Monad.Free
   ( Free
-  , suspendF
-  , wrap
-  , liftF
-  , hoistFree
-  , foldFree
-  , substFree
-  , runFree
-  , runFreeM
+  , lift
+  , roll
+  , suspend
+  , hoist
   , resume
-  , resume'
-  ) where
+  , run
+  , runRec
+  , runPure
+  , interpret
+  , interpretRec
+  )
+  where
 
 import Prelude
 
-import Control.Apply (lift2)
 import Control.Monad.Rec.Class (class MonadRec, Step(..), tailRecM)
-import Control.Monad.Trans.Class (class MonadTrans)
-
-import Data.CatList (CatList, empty, snoc, uncons)
-import Data.Either (Either(..))
-import Data.Eq (class Eq1, eq1)
-import Data.Foldable (class Foldable, foldMap, foldl, foldr)
-import Data.Maybe (Maybe(..))
-import Data.Ord (class Ord1, compare1)
-import Data.Traversable (class Traversable, traverse)
-import Data.Tuple (Tuple(..))
-
+import Data.Exists (Exists, mkExists, runExists)
 import Unsafe.Coerce (unsafeCoerce)
 
--- | The free monad for a type constructor `f`.
--- |
--- | Implemented in the spirit of [Reflection without Remorse](http://okmij.org/ftp/Haskell/zseq.pdf),
--- | the free monad is represented using a sequential data structure in
--- | order to overcome the quadratic complexity of left-associated binds
--- | and traversal through the free monad structure.
-data Free f a = Free (FreeView f Val Val) (CatList (ExpF f))
+foreign import data UnsafeBoundValue :: Type
 
-newtype ExpF f = ExpF (Val -> Free f Val)
+foreign import data UnsafeBoundF :: Type -> Type
 
-data FreeView f a b = Return a | Bind (f b) (b -> Free f a)
+data Free f a
+  = Pure a
+  | Bind (f UnsafeBoundValue) (FreeBinds f UnsafeBoundValue a)
 
-data Val
+data FreeView f a b
+  = PureView a
+  | BindView (f b) (b -> Free f a)
 
-instance eqFree :: (Functor f, Eq1 f, Eq a) => Eq (Free f a) where
-  eq x y = case resume x, resume y of
-    Left fa, Left fb -> eq1 fa fb
-    Right a, Right b -> a == b
-    _, _ -> false
+data FreeBinds f a b
+  = Leaf (a -> Free f b)
+  | Node (FreeBinds f a UnsafeBoundValue) (FreeBinds f UnsafeBoundValue b)
+  | Hoist (UnsafeBoundF ~> f) (FreeBinds UnsafeBoundF a b)
 
-instance eq1Free :: (Functor f, Eq1 f) => Eq1 (Free f) where
-  eq1 = eq
+data FreeCons f a b
+  = FreeCons (a -> Free f UnsafeBoundValue) (FreeBinds f UnsafeBoundValue b)
 
-instance ordFree :: (Functor f, Ord1 f, Ord a) => Ord (Free f a) where
-  compare x y = case resume x, resume y of
-    Left fa, Left fb -> compare1 fa fb
-    Left _, _ -> LT
-    _, Left _ -> GT
-    Right a, Right b -> compare a b
+lift :: forall f a. f a -> Free f a
```

Comment:
I don't like it
1) Breaking changes
2) `lift ` is standard name for `MonadTrans`

Reply:
I'm open to discussing whether `liftF` should remain the name of this function. It's not _only_ used for `MonadTrans` (it's also used for say Halogen's lift, or `run`'s lift) but if folks would really like it to remain `liftF` I can see why.

Do you have any other comments for this PR?